### PR TITLE
test: adds unit tests for Reconcile

### LINKED ISSUE #44: Add unit test for status condition update when has is not changed
Ref - https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/36#discussion_r1678742548

/assign

### PR DESCRIPTION
**Pull request by nilekhc**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, using `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when the PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/35
Fixes #44 

**Special notes for your reviewer**:

**TODOs**:

- [x] squashed commits
- [ ] includes documentation
- [x] adds unit tests

/label tide/merge-method-squash

### CODE REVIEW

**Comment by nilekhc** [comment]
> @nilekhc: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-secrets-store-sync-controller-image-scan	[701d0a7](https://github.com/kubernetes-sigs/secrets-store-sync-controller/commit/701d0a79b00bde08ada24be149fe3f876fbeb366)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_secrets-store-sync-controller/36/pull-secrets-store-sync-controller-image-scan/1810402122299084800)	false	`/test pull-secrets-store-sync-controller-image-scan`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes-sigs&repo=secrets-store-sync-controller&pr=36). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Anilekhc). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues?q=is:issue+is:open) when you hit one in your PR.

This will be fixed with https://github.com/kubernetes-sigs/secrets-store-sync-controller/pull/30

File: controllers/secretsync_controller_test.go [hunk 1]
```diff
@@ -91,7 +91,298 @@ func TestReconcile(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "sse2esecret",
+					Namespace: "default",
+				},
+				Data: map[string][]byte{
+					"foo": []byte("bar"),
+				},
+			},
+			expectedError:       true,
+			expectedErrorString: "secretsyncs.secret-sync.x-k8s.io \"sse2esecret\" not found",
```
Comment (aramase) [inline controllers/secretsync_controller_test.go:121]:
```suggestion
			expectedErrorString: `secretsyncs.secret-sync.x-k8s.io "sse2esecret" not found`,
```

File: controllers/secretsync_controller_test.go [hunk 2]
```diff
@@ -91,7 +91,298 @@ func TestReconcile(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "sse2esecret",
+					Namespace: "default",
+				},
+				Data: map[string][]byte{
+					"foo": []byte("bar"),
+				},
+			},
+			expectedError:       true,
+			expectedErrorString: "secretproviderclasses.secrets-store.csi.x-k8s.io \"test-spc\" not found",
```
Comment (aramase) [inline controllers/secretsync_controller_test.go:295]:
```suggestion
			expectedErrorString: `secretproviderclasses.secrets-store.csi.x-k8s.io "test-spc" not found`,
```

File: controllers/secretsync_controller_test.go [hunk 3]
```diff
@@ -91,7 +91,298 @@ func TestReconcile(t *testing.T) {
...
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "sse2esecret",
+					Namespace: "default",
+				},
+				Data: map[string][]byte{
+					"foo": []byte("bar"),
+				},
+			},
+			expectedError:       true,
+			expectedErrorString: "provider not found: provider \"invalid-fake-provider\"",
```
Comment (aramase) [inline controllers/secretsync_controller_test.go:340]:
```suggestion
			expectedErrorString: `provider not found: provider "invalid-fake-provider"`,
```

File: controllers/secretsync_controller_test.go [hunk 4]
```diff
@@ -110,8 +401,15 @@ func TestReconcile(t *testing.T) {
...
+			if test.expectedError {
+				if err == nil {
+					t.Fatalf("expected error but got nil")
+				}
+				if err.Error() != test.expectedErrorString {
+					t.Fatalf("expected error %s but got %s", test.expectedErrorString, err.Error())
+				}
+			} else if err != nil {
+				t.Fatalf("did not expect an error but got: %v", err)
 			}
```
Comment (aramase) [inline controllers/secretsync_controller_test.go:413]:
```suggestion
			if len(test.expectedErrorString) > 0 {
				if err == nil || err.Error() != test.expectedErrorString {
					t.Fatalf("expected error %q, got %q", test.expectedErrorString, err)
				}
			} else if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
```

We don't need 2 variables (bool and string) to define the required error cases.

File: pkg/util/secretutil/secret.go
Comment (aramase) [inline pkg/util/secretutil/secret.go:1]: This file needs unit tests. Can be done in a follow-up.

File: controllers/secretsync_controller_test.go [hunk 5]
```diff
@@ -110,8 +401,15 @@ func TestReconcile(t *testing.T) {
...
 
 			_, err := ssc.Reconcile(context.Background(), req)
```
Comment (aramase) [inline controllers/secretsync_controller_test.go:403]: Can we validate the conditions being set?

File: controllers/secretsync_controller_test.go [hunk 6]
```diff
@@ -91,7 +91,298 @@ func TestReconcile(t *testing.T) {
...
+					Name:      "sse2esecret",
+					Namespace: "default",
+				},
+				Data: map[string][]byte{
+					"foo": []byte("bar"),
+				},
+			},
+			expectedError:       true,
+			expectedErrorString: "key in secretObject.data is empty",
 		},
```
Comment (aramase) [inline controllers/secretsync_controller_test.go:386]: Need unit tests for

https://github.com/kubernetes-sigs/secrets-store-sync-controller/blob/569742b2ec0ff963a19ccc94b93c209197c03970/controllers/secretsync_controller.go#L242-L245

**Review by aramase — CHANGES_REQUESTED** [review]
Added few comments

File: pkg/util/secretutil/secret.go
Comment (nilekhc) [reply pkg/util/secretutil/secret.go:1]: https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/43

File: controllers/secretsync_controller_test.go (hunk 6 above)
Comment (nilekhc) [reply controllers/secretsync_controller_test.go]: https://github.com/kubernetes-sigs/secrets-store-sync-controller/issues/44

Comment (nilekhc) [reply controllers/secretsync_controller_test.go]: I added test in this PR itself

**Review by aramase — COMMENTED** [review]
/lgtm
/approve

**Comment by aramase** [comment]
/lgtm cancel

The unit tests are failing.

**Comment by nilekhc** [comment]
> /lgtm cancel
> 
> The unit tests are failing.

@aramase test is now passing

**Review by aramase — COMMENTED** [review]
/lgtm

**Pull request merged**